packages/x-utils/src/types/object.types.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * Picks the keys from the provided object that match the given type.
+ *
+ * @param SomeObject - The object to pick the keys from.
+ * @param Type - The type of the properties to pick.
+ *
+ * @example
+ * ```typescript
+ * interface Result {
+ *   id: string,
+ *   price: {
+ *     max: number,
+ *     min: number,
+ *     symbol: string;
+ *   },
+ *   images: string[]
+ * }
+ *
+ * type ResultOnlyStrings = PickByType<Result, string>; // { id: string; price: { symbol: string; }}
+ * type ResultOnlyNumbers = PickByType<Result, number>; // { price: { max: number; min: number; }}
+ * ```
+ *
+ * @public
+ */
+export type PickByType<SomeObject, Type> = {
```

Comment:
Renamed `TreeShakeObject` to `PickByType`, which should be more straightforward to understand. Removed the `ExcludeOptionals` generic as @JaviIG suggested, as you can always pass the union of your desired `type` and `undefined` (i.e. `string | undefined`). Changed `object` to `Record<string, unknown>` as @JaviIG pointed out `functions` and `arrays` were being identified as `object` and therefore also being included.